### PR TITLE
Make storage providers configurable

### PR DESCRIPTION
**Pull request by erikjohnston**
I'm not the worlds biggest fan when it comes to how we store the class in the config, but it matches what we do with auth providers.

### CODE REVIEW

File: synapse/config/repository.py [hunk 1]
```diff
@@ -73,16 +79,65 @@ def read_config(self, config):
...
+            backup_media_store_path = self.ensure_directory(
                 self.backup_media_store_path
             )
 
-        self.synchronous_backup_media_store = config.get(
+        synchronous_backup_media_store = config.get(
             "synchronous_backup_media_store", False
         )
 
+        storage_providers = config.get("media_storage_providers", [])
```
Comment (richvdh) [inline synapse/config/repository.py]: move this to line 81, then you only need one `if backup_media_store_path:`

File: synapse/config/repository.py [hunk 2]
```diff
@@ -73,16 +79,65 @@ def read_config(self, config):
 
         self.media_store_path = self.ensure_directory(config["media_store_path"])
 
-        self.backup_media_store_path = config.get("backup_media_store_path")
-        if self.backup_media_store_path:
-            self.backup_media_store_path = self.ensure_directory(
+        backup_media_store_path = config.get("backup_media_store_path")
+        if backup_media_store_path:
+            backup_media_store_path = self.ensure_directory(
```
Comment (richvdh) [inline synapse/config/repository.py]: this should be after the test on `storage_providers`

File: synapse/config/repository.py [hunk 3]
```diff
@@ -36,6 +38,10 @@
     "ThumbnailRequirement", ["width", "height", "method", "media_type"]
 )
 
+MediaStorageProviderConfig = namedtuple(
```
Comment (richvdh) [inline synapse/config/repository.py]: can we document the members on here somehow, please?

File: synapse/rest/media/v1/storage_provider.py [hunk 4]
```diff
@@ -97,13 +98,13 @@ class FileStorageProviderBackend(StorageProvider):
     """A storage provider that stores files in a directory on a filesystem.
 
     Args:
-        cache_directory (str): Base path of the local media repository
-        base_directory (str): Base path to store new files
+        hs (HomeServer)
+        config: The config returned by `parse_config`, i
```
Comment (richvdh) [inline synapse/rest/media/v1/storage_provider.py]: i

File: synapse/rest/media/v1/storage_provider.py [hunk 5]
```diff
@@ -125,3 +126,14 @@ def fetch(self, path, file_info):
         backup_fname = os.path.join(self.base_directory, path)
         if os.path.isfile(backup_fname):
             return FileResponder(open(backup_fname, "rb"))
+
+    def parse_config(config):
```
Comment (richvdh) [inline synapse/rest/media/v1/storage_provider.py]: needs an `@staticmethod`

File: synapse/rest/media/v1/storage_provider.py [hunk 6]
```diff
@@ -125,3 +126,14 @@ def fetch(self, path, file_info):
...
+        """Called on startup to parse config supplied. This should parse
+        the config and raise if there is a problem.
+
+        The returned value is passed into the constructor.
+
+        In this case we only care about a single param, the directory, so lets
```
Comment (richvdh) [inline synapse/rest/media/v1/storage_provider.py]: let's

File: synapse/config/repository.py (hunk 2 above)
Comment (richvdh) [reply synapse/config/repository.py]: is it even needed given FileStorageProviderBackend does it?

File: synapse/config/repository.py (hunk 3 above)
Comment (richvdh) [reply synapse/config/repository.py]: ok thanks. for future ref, I'd like to see their types made explicit too, but this is fine for now

Comment (erikjohnston) [reply synapse/config/repository.py]: Ah, yup, oops.

**Pull request merged**